src/Bicep.RegistryModuleTool/ModuleFiles/MainArmTemplateParametersFile.cs

```diff
@@ -0,0 +1,151 @@
+                    writer.WriteStartObject();
+                    writer.WritePropertyName("value");
+                    JsonElementFactory.CreateElement(parameterInstance.Value).WriteTo(writer);
+                    writer.WriteEndObject();
+                }
+
+                writer.WriteEndObject();
+
+                writer.WriteEndObject();
+            }
+
+            var path = fileSystem.Path.GetFullPath(FileName);
+            var content = Encoding.UTF8.GetString(bufferWriter.WrittenSpan);
+            var rootElement = JsonElementFactory.CreateElement(bufferWriter.WrittenMemory);
+
+            return new(path, content, rootElement);
+        }
+
+        public static MainArmTemplateParametersFile ReadFromFileSystem(IFileSystem fileSystem)
+        {
+            var path = fileSystem.Path.GetFullPath(FileName);
+
+            try
+            {
+                var content = fileSystem.File.ReadAllText(path);
+                var jsonElement = JsonElementFactory.CreateElement(content);
+
+                return new(path, content, jsonElement);
+            }
+            catch (JsonException jsonException)
+            {
+                throw new BicepException($"The ARM template parameters file \"{path}\" is not a valid JSON file. {jsonException.Message}");
+            }
+        }
+
+        public void WriteToFileSystem(IFileSystem fileSystem) => fileSystem.File.WriteAllText(FileName, this.Content);
+
+        protected override void ValidatedBy(IModuleFileValidator validator) => validator.Validate(this);
+
+        private static IEnumerable<MainArmTemplateParameterInstance> GenerateParameterInstancesWithExistingFile(
+            MainArmTemplateParametersFile existingFile,
+            MainArmTemplateFile mainArmTemplateFile)
+        {
+            var mergedParameterInstances = new List<MainArmTemplateParameterInstance>();
+            var parametersByName = mainArmTemplateFile.Parameters.ToDictionary(x => x.Name, x => x);
+            var parameterInstancesByName = existingFile.ParameterInstances.ToDictionary(x => x.Name, x => x);
+
+            foreach (var parameter in mainArmTemplateFile.Parameters)
+            {
+                if (parameterInstancesByName.TryGetValue(parameter.Name, out var existingInstance))
+                {
+                    mergedParameterInstances.Add(existingInstance);
+                }
+                else if (parameter.Required)
+                {
+                    mergedParameterInstances.Add(GenerateDummyParameterInstance(parameter));
+                }
+            }
+
+            return mergedParameterInstances;
+        }
+
+        private static MainArmTemplateParameterInstance GenerateDummyParameterInstance(MainArmTemplateParameter parameter)
+        {
+            var name = parameter.Name;
+            object value = parameter.Type.ToLowerInvariant() switch
+            {
+                "bool" => false,
```

Comment:
>false [](http://example.com/codeflow?start=26&length=5)

Should we generate intentionally broken values? Maybe always a string "TODO" to force the user to fix it?

Reply:
That's a good point! As a "TODO" string may still be a valid parameter value, I feel like it may not be enough to force the user to fix it, so I changed it to generate an empty object `{}` for each parameter, which will fail the schema validation for sure. What do you think?

Reply:
`{}` will be valid for object params, but object params. TODO has the benefit of being self-descriptive, but most params are strings, so it definitely won't force the user to update it.

I can't think of a good way to generate malformed JSON without doing our own JSON serialization, which seems like too much.

Random idea: We could generate a string placeholder that's unlikely to be placed in a param. Something like "--< TODO >--". Then validation could look for it and fail the build. The extra prefix and suffix would allow the user to still have TODO as a value without issues.